File: app/src/main/java/com/genonbeta/TrebleShot/graphics/drawable/TextDrawable.java
package com.genonbeta.TrebleShot.graphics.drawable;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.OvalShape;
import android.graphics.drawable.shapes.RectShape;
import android.graphics.drawable.shapes.RoundRectShape;

import com.genonbeta.TrebleShot.util.TextUtils;

public class TextDrawable extends ShapeDrawable
{
    private static final float SHADE_FACTOR = 0.9f;

    private final Paint mTextPaint;
    private final Paint mBorderPaint;
    private final String mText;
    private final RectShape mShape;
    private final int mColor;
    private final int mHeight;
    private final int mWidth;
    private final int mFontSize;
    private final int mBorderThickness;
    private final float mRadius;

    private TextDrawable(Builder builder)
    {
        super(builder.mShape);

        // shape properties
        mShape = builder.mShape;
        mHeight = builder.mHeight;
        mWidth = builder.mWidth;
        mRadius = builder.mRadius;

        // text and color
        String processedText = builder.mToUpperCase ? builder.mText.toUpperCase() : builder.mText;

        if (builder.mTextMaxLength >= 0) {
            processedText = builder.mFirstLetters
                    ? TextUtils.getLetters(processedText, builder.mTextMaxLength)
                    : (processedText.length() > builder.mTextMaxLength) ? processedText.substring(0, builder.mTextMaxLength) : processedText;
        }

        mText = processedText;
        mColor = builder.mColor;

        // text paint settings
        mFontSize = builder.mFontSize;
        mTextPaint = new Paint();
        mTextPaint.setColor(builder.mTextColor);
        mTextPaint.setAntiAlias(true);
        mTextPaint.setFakeBoldText(builder.mIsBold);
        mTextPaint.setStyle(Paint.Style.FILL);
        mTextPaint.setTypeface(builder.mFont);
        mTextPaint.setTextAlign(Paint.Align.CENTER);
        mTextPaint.setStrokeWidth(builder.mBorderThickness);

        // border paint settings
        mBorderThickness = builder.mBorderThickness;
        mBorderPaint = new Paint();
        mBorderPaint.setColor(getDarkerShade(mColor));
        mBorderPaint.setStyle(Paint.Style.STROKE);
        mBorderPaint.setStrokeWidth(mBorderThickness);

        // drawable paint mColor
        Paint paint = getPaint();
        paint.setColor(mColor);
    }

    public static IShapeBuilder builder()
    {
        return new Builder();
    }

    private int getDarkerShade(int color)
    {
        return Color.rgb((int) (SHADE_FACTOR * Color.red(color)),
                (int) (SHADE_FACTOR * Color.green(color)),
                (int) (SHADE_FACTOR * Color.blue(color)));
    }

    @Override
    public void draw(Canvas canvas)
    {
        super.draw(canvas);
        Rect r = getBounds();

        // draw border
        if (mBorderThickness > 0)
            drawBorder(canvas);

        int count = canvas.save();
        canvas.translate(r.left, r.top);

        // draw text
        int width = mWidth < 0 ? r.width() : mWidth;
        int height = mHeight < 0 ? r.height() : mHeight;
        int fontSize = mFontSize < 0 ? (Math.min(width, height) / 2) : mFontSize;

        mTextPaint.setTextSize(fontSize);

        canvas.drawText(mText, width / 2, height / 2 - ((mTextPaint.descent() + mTextPaint.ascent()) / 2), mTextPaint);
        canvas.restoreToCount(count);
    }

    private void drawBorder(Canvas canvas)
    {
        RectF rect = new RectF(getBounds());
        rect.inset(mBorderThickness / 2, mBorderThickness / 2);

        if (mShape instanceof OvalShape) {
            canvas.drawOval(rect, mBorderPaint);
        } else if (mShape instanceof RoundRectShape) {
            canvas.drawRoundRect(rect, mRadius, mRadius, mBorderPaint);
        } else {
            canvas.drawRect(rect, mBorderPaint);
        }
    }

    @Override
    public void setAlpha(int alpha)
    {
        mTextPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter cf)
    {
        mTextPaint.setColorFilter(cf);
    }

    @Override
    public int getOpacity()
    {
        return PixelFormat.TRANSLUCENT;
    }

    @Override
    public int getIntrinsicWidth()
    {
        return mWidth;
    }

    @Override
    public int getIntrinsicHeight()
    {
        return mHeight;
    }

    public interface IConfigBuilder
    {
        IConfigBuilder bold();

        IConfigBuilder firstLettersOnly(boolean use);

        IConfigBuilder fontSize(int size);

        IConfigBuilder height(int height);

        IConfigBuilder shapeColor(int color);

        IConfigBuilder textColor(int color);

        IConfigBuilder textMaxLength(int length);

        IConfigBuilder toUpperCase();

        IConfigBuilder useFont(Typeface font);

        IConfigBuilder width(int width);

        IConfigBuilder withBorder(int thickness);

        IShapeBuilder endConfig();
    }

    public interface IBuilder
    {
        TextDrawable build(String text);
    }

    public interface IShapeBuilder
    {
        IConfigBuilder beginConfig();

        IBuilder rect();

        IBuilder round();

        IBuilder roundRect(int radius);

        TextDrawable buildRect(String text);

        TextDrawable buildRound(String text);

        TextDrawable buildRoundRect(String text, int radius);
    }

    public static class Builder implements IConfigBuilder, IShapeBuilder, IBuilder
    {
        private int mColor;
        private int mBorderThickness;
        private int mWidth;
        private int mHeight;
        private int mTextColor;
        private int mTextMaxLength;
        private int mFontSize;
        private float mRadius;
        private boolean mToUpperCase;
        private boolean mIsBold;
        private boolean mFirstLetters;
        private String mText;
        private Typeface mFont;
        private RectShape mShape;

        private Builder()
        {
            mText = "";
            mColor = Color.GRAY;
            mTextColor = Color.WHITE;
            mBorderThickness = 0;
            mWidth = -1;
            mHeight = -1;
            mShape = new RectShape();
            mFont = Typeface.create("sans-serif", Typeface.NORMAL);
            mTextMaxLength = -1;
            mFontSize = -1;
            mIsBold = false;
            mToUpperCase = false;
        }

        public IConfigBuilder width(int width)
        {
            mWidth = width;
            return this;
        }

        public IConfigBuilder height(int height)
        {
            mHeight = height;
            return this;
        }

        @Override
        public IConfigBuilder shapeColor(int color)
        {
            mColor = color;
            return this;
        }

        public IConfigBuilder textColor(int color)
        {
            mTextColor = color;
            return this;
        }

        @Override
        public IConfigBuilder textMaxLength(int length)
        {
            mTextMaxLength = length;
            return this;
        }

        public IConfigBuilder withBorder(int thickness)
        {
            mBorderThickness = thickness;
            return this;
        }

        public IConfigBuilder useFont(Typeface font)
        {
            mFont = font;
            return this;
        }

        public IConfigBuilder fontSize(int size)
        {
            mFontSize = size;
            return this;
        }

        public IConfigBuilder bold()
        {
            mIsBold = true;
            return this;
        }

        public IConfigBuilder toUpperCase()
        {
            mToUpperCase = true;
            return this;
        }

        @Override
        public IConfigBuilder firstLettersOnly(boolean use)
        {
            mFirstLetters = use;
            return this;
        }

        @Override
        public IConfigBuilder beginConfig()
        {
            return this;
        }

        @Override
        public IShapeBuilder endConfig()
        {
            return this;
        }

        @Override
        public IBuilder rect()
        {
            mShape = new RectShape();
            return this;
        }

        @Override
        public IBuilder round()
        {
            mShape = new OvalShape();
            return this;
        }

        @Override
        public IBuilder roundRect(int radius)
        {
            mRadius = radius;
            float[] radii = {radius, radius, radius, radius, radius, radius, radius, radius};
            mShape = new RoundRectShape(radii, null, null);
            return this;
        }

        @Override
        public TextDrawable buildRect(String text)
        {
            rect();
            return build(text);
        }

        @Override
        public TextDrawable buildRoundRect(String text, int radius)
        {
            roundRect(radius);
            return build(text);
        }

        @Override
        public TextDrawable buildRound(String text)
        {
            round();
            return build(text);
        }

        @Override
        public TextDrawable build(String text)
        {
            mText = text;
            return new TextDrawable(this);
        }
    }
}

File: app/src/main/java/com/genonbeta/TrebleShot/graphics/drawable/TextDrawable.kt
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.graphics.drawable

import android.graphics.*
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.graphics.drawable.shapes.RectShape
import android.graphics.drawable.shapes.RoundRectShape
import com.genonbeta.TrebleShot.util.TextUtils

class TextDrawable private constructor(builder: Builder) : ShapeDrawable(builder.mShape) {
    private val mTextPaint: Paint
    private val mBorderPaint: Paint
    private val mText: String
    private val mShape: RectShape
    private val mColor: Int
    private val mHeight: Int
    private val mWidth: Int
    private val mFontSize: Int
    private val mBorderThickness: Int
    private val mRadius: Float
    private fun getDarkerShade(color: Int): Int {
        return Color.rgb(
            (SHADE_FACTOR * Color.red(color)).toInt(),
            (SHADE_FACTOR * Color.green(color)).toInt(),
            (SHADE_FACTOR * Color.blue(color)).toInt()
        )
    }

    override fun draw(canvas: Canvas) {
        super.draw(canvas)
        val r: Rect = bounds

        // draw border
        if (mBorderThickness > 0) drawBorder(canvas)
        val count = canvas.save()
        canvas.translate(r.left.toFloat(), r.top.toFloat())

        // draw text
        val width = if (mWidth < 0) r.width() else mWidth
        val height = if (mHeight < 0) r.height() else mHeight
        val fontSize = if (mFontSize < 0) Math.min(width, height) / 2 else mFontSize
        mTextPaint.textSize = fontSize.toFloat()
        canvas.drawText(
            mText,
            (width / 2).toFloat(),
            height / 2 - (mTextPaint.descent() + mTextPaint.ascent()) / 2,
            mTextPaint
        )
        canvas.restoreToCount(count)
    }

    private fun drawBorder(canvas: Canvas) {
        val rect = RectF(getBounds())
        rect.inset((mBorderThickness / 2).toFloat(), (mBorderThickness / 2).toFloat())
        if (mShape is OvalShape) {
            canvas.drawOval(rect, mBorderPaint)
        } else if (mShape is RoundRectShape) {
            canvas.drawRoundRect(rect, mRadius, mRadius, mBorderPaint)
        } else {
            canvas.drawRect(rect, mBorderPaint)
        }
    }

    override fun setAlpha(alpha: Int) {
        mTextPaint.alpha = alpha
    }

    override fun setColorFilter(cf: ColorFilter?) {
        mTextPaint.colorFilter = cf
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun getIntrinsicWidth(): Int {
        return mWidth
    }

    override fun getIntrinsicHeight(): Int {
        return mHeight
    }

    interface IConfigBuilder {
        fun bold(): IConfigBuilder
        fun firstLettersOnly(use: Boolean): IConfigBuilder
        fun fontSize(size: Int): IConfigBuilder
        fun height(height: Int): IConfigBuilder
        fun shapeColor(color: Int): IConfigBuilder
        fun textColor(color: Int): IConfigBuilder
        fun textMaxLength(length: Int): IConfigBuilder
        fun toUpperCase(): IConfigBuilder
        fun useFont(font: Typeface): IConfigBuilder
        fun width(width: Int): IConfigBuilder
        fun withBorder(thickness: Int): IConfigBuilder
        fun endConfig(): IShapeBuilder
    }

    interface IBuilder {
        fun build(text: String): TextDrawable
    }

    interface IShapeBuilder {
        fun beginConfig(): IConfigBuilder
        fun rect(): IBuilder
        fun round(): IBuilder
        fun roundRect(radius: Int): IBuilder
        fun buildRect(text: String): TextDrawable
        fun buildRound(text: String): TextDrawable
        fun buildRoundRect(text: String, radius: Int): TextDrawable
    }

    class Builder : IConfigBuilder, IShapeBuilder, IBuilder {
        var mColor: Int
        var mBorderThickness: Int
        var mWidth: Int
        var mHeight: Int
        var mTextColor: Int
        var mTextMaxLength: Int
        var mFontSize: Int
        var mRadius = 0f
        var mToUpperCase: Boolean
        var mIsBold: Boolean
        var mFirstLetters = false
        var mText = ""
        var mFont: Typeface
        var mShape: RectShape
        override fun width(width: Int): IConfigBuilder {
            mWidth = width
            return this
        }

        override fun height(height: Int): IConfigBuilder {
            mHeight = height
            return this
        }

        override fun shapeColor(color: Int): IConfigBuilder {
            mColor = color
            return this
        }

        override fun textColor(color: Int): IConfigBuilder {
            mTextColor = color
            return this
        }

        override fun textMaxLength(length: Int): IConfigBuilder {
            mTextMaxLength = length
            return this
        }

        override fun withBorder(thickness: Int): IConfigBuilder {
            mBorderThickness = thickness
            return this
        }

        override fun useFont(font: Typeface): IConfigBuilder {
            mFont = font
            return this
        }

        override fun fontSize(size: Int): IConfigBuilder {
            mFontSize = size
            return this
        }

        override fun bold(): IConfigBuilder {
            mIsBold = true
            return this
        }

        override fun toUpperCase(): IConfigBuilder {
            mToUpperCase = true
            return this
        }

        override fun firstLettersOnly(use: Boolean): IConfigBuilder {
            mFirstLetters = use
            return this
        }

        override fun beginConfig(): IConfigBuilder {
            return this
        }

        override fun endConfig(): IShapeBuilder {
            return this
        }

        override fun rect(): IBuilder {
            mShape = RectShape()
            return this
        }

        override fun round(): IBuilder {
            mShape = OvalShape()
            return this
        }

        override fun roundRect(radius: Int): IBuilder {
            mRadius = radius.toFloat()
            val radii = floatArrayOf(
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat(),
                radius.toFloat()
            )
            mShape = RoundRectShape(radii, null, null)
            return this
        }

        override fun buildRect(text: String): TextDrawable {
            rect()
            return build(text)
        }

        override fun buildRoundRect(text: String, radius: Int): TextDrawable {
            roundRect(radius)
            return build(text)
        }

        override fun buildRound(text: String): TextDrawable {
            round()
            return build(text)
        }

        override fun build(text: String): TextDrawable {
            mText = text
            return TextDrawable(this)
        }

        init {
            mColor = Color.GRAY
            mTextColor = Color.WHITE
            mBorderThickness = 0
            mWidth = -1
            mHeight = -1
            mShape = RectShape()
            mFont = Typeface.create("sans-serif", Typeface.NORMAL)
            mTextMaxLength = -1
            mFontSize = -1
            mIsBold = false
            mToUpperCase = false
        }
    }

    companion object {
        private const val SHADE_FACTOR = 0.9f
        fun builder(): IShapeBuilder {
            return Builder()
        }
    }

    init {

        // shape properties
        mShape = builder.mShape
        mHeight = builder.mHeight
        mWidth = builder.mWidth
        mRadius = builder.mRadius

        // text and color
        var processedText: String = if (builder.mToUpperCase) builder.mText.toUpperCase() else builder.mText
        if (builder.mTextMaxLength >= 0) {
            processedText = when {
                builder.mFirstLetters -> TextUtils.getLetters(processedText, builder.mTextMaxLength)
                processedText.length > builder.mTextMaxLength -> processedText.substring(0, builder.mTextMaxLength)
                else -> processedText
            }
        }
        mText = processedText
        mColor = builder.mColor

        // text paint settings
        mFontSize = builder.mFontSize
        mTextPaint = Paint()
        mTextPaint.color = builder.mTextColor
        mTextPaint.isAntiAlias = true
        mTextPaint.isFakeBoldText = builder.mIsBold
        mTextPaint.style = Paint.Style.FILL
        mTextPaint.typeface = builder.mFont
        mTextPaint.textAlign = Paint.Align.CENTER
        mTextPaint.strokeWidth = builder.mBorderThickness.toFloat()

        // border paint settings
        mBorderThickness = builder.mBorderThickness
        mBorderPaint = Paint()
        mBorderPaint.color = getDarkerShade(mColor)
        mBorderPaint.style = Paint.Style.STROKE
        mBorderPaint.strokeWidth = mBorderThickness.toFloat()

        // drawable paint mColor
        val paint: Paint = getPaint()
        paint.color = mColor
    }
}